propolis-server/src/api.rs

```diff
@@ -0,0 +1,189 @@
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct InstancePathParams {
+    pub instance_id: Uuid,
+}
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct InstanceEnsureRequest {
+    pub properties: InstanceProperties,
+}
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct InstanceEnsureResponse {}
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct InstanceGetResponse {
+    pub instance: Instance,
+}
+
+/// Requested state of an Instance.
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct InstanceStateChange {
+    pub state: InstanceStateRequested,
+}
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub enum InstanceStateRequested {
+    Run,
+    Stop,
+    Reboot,
+}
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct InstanceSerialRequest {
+    pub bytes: Vec<u8>,
+}
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct InstanceSerialResponse {
+    pub bytes: Vec<u8>,
+}
+
+/// Current state of an Instance.
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub enum InstanceState {
+    Creating,
+    Starting,
+    Running,
+    Stopping,
+    Stopped,
+    Repairing,
+    Failed,
+    Destroyed,
+}
+
+#[derive(Clone, Deserialize, Serialize, JsonSchema)]
+pub struct InstanceProperties {
+    pub generation_id: u64,
+
+    /// Unique identifier for this Instance.
+    pub id: Uuid,
+    /// Human-readable name of the Instance.
+    pub name: String,
+    /// Free-form text description of an Instance.
+    pub description: String,
+
+    /// ID of the image used to initialize this Instance.
+    pub image_id: Uuid,
+    /// ID of the bootrom used to initialize this Instance.
+    pub bootrom_id: Uuid,
+    /// Size of memory allocated to the Instance, in MB.
+    pub memory: u64,
+    /// Number of vCPUs to be allocated to the Instance.
+    pub vcpus: u8,
+}
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct Instance {
+    pub properties: InstanceProperties,
+    pub state: InstanceState,
+
+    pub disks: Vec<DiskAttachment>,
+    pub nics: Vec<NicAttachment>,
+}
```

Comment:
This portion is being "used", so to speak.

The rest is copied from RFD 152, but still very much in flux.